[occm] Get IP addresses of neutron subports

jingczhang opened the pull request:
Vlan-aware VMs are commonly used in telecom.
Those VMs are plugged into flat networks and use neutron trunk.
https://docs.openstack.org/neutron/latest/admin/config-trunking.html

Currently, getAttachedInterfacesByID() only returns neutron ports directly attached to VM.
For vlan-aware VMs, IP addresses are assigned on neutron subports.
Subports are attached to the trunk; they are not attached to the VM directly.
This pull request improves getAttachedInterfacesByID() to return IP addresses of neutron subports when they exist.
Without this change, Kubernetes is unable to get IP addresses of vlan-aware VMs.
This change is transparent to VMs not using neutron trunk.

**Special notes for reviewers**:
This is a re-attempt of the following PR which was idle out:
https://github.com/kubernetes/cloud-provider-openstack/pull/1535
The code has been in production for two years now.

**Release note**:
<!--
1. Release note is required if a significant change is introduced, otherwise please keep this section as is.
2. Release note is in Markdown format and should begin with the binary name unless multiple binaries are affected, e.g. [openstack-cloud-controller-manager] Deprecate Neutron-LBaaS support.
3. Instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

jichenjc commented:
/ok-to-test

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +721,91 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+	// Get subports attached to the trunk
+	var subports []trunks.Subport
+	for _, trunk := range allTrunks {
+		for _, iface := range interfaces {
```
Interfaces is still empty? That would make the allTrunks loop a no-op.

jingczhang replied:
Sharp eyes! My mistake when creating the PR, a rebase error, the functional prototype should be the following, I will correct it.
// getSubInterfaces
func getSubInterfaces(interfaces []attachinterfaces.Interface, network *gophercloud.ServiceClient) ([]attachinterfaces.Interface, error)

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +721,91 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+	klog.V(5).Infof("subports %v", subports)
+
+	// Arrange subports as for directly attached ports
+	for _, sport := range subports {
```
Given that the previous loop looks like a no-op I think subports is guaranteed to be empty here.

jingczhang replied:
Acked, will handle with the above comment.

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +721,91 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+			klog.Errorf("Failed to get port info for subport %s: %v", sport.PortID, err)
+			return interfaces, err
+		}
+		n, err := networks.Get(network, p.NetworkID).Extract()
```
A network query per subport doesn't seem like a good idea.

jingczhang replied:
Need to get the network name for filtering on internal-network-name later on.
https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md#networking

jingczhang replied:
Point taken, we will keep this mind when working on PR update.

dulek replied:
We could probably list networks and then use that list to find needed info. It has downsides too.

mdbooth reviewed (commented):
Apologies if I've missed something really obvious, but this doesn't look right to me. I can't see how `getSubInterfaces()` in particular can work. Is there some code missing which populates interfaces? Even If it did work it would have potentially catastrophic performance impacts on some clouds due to the nested API calls. I don't think we should merge this with its current code structure.

I wonder if we can do something much more efficient with https://docs.openstack.org/api-ref/network/v2/index.html?expanded=list-trunks-detail,show-trunk-details-detail#show-trunk-details?

@dulek Can you take a look at this when you're around?

stephenfin commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +721,91 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+	listOpts := trunks.ListOpts{}
+	allPages, err := trunks.List(network, listOpts).AllPages()
+	if err != nil {
+		klog.Errorf("Failed to list trunks: %v", err)
```
You need to check for this extension, ideally once at startup. I don't know if gophercloud does this for us. I imagine not.

jingczhang replied:
Neutron trunk had been existing for many releases, before CCM? I would expect API error, Openstack instances we have including N-k (k is big) all support neutron trunk, let alone the Openstack instances running Kubernetes.

jichenjc replied:
sorry, I mean some openstack didn't have this extension ... it's common that some plugins are not enabled

jingczhang replied:
Good point, we will check and ensure code change is transparent when trunk plugin is not configured.

jingczhang replied:
We have checked a deployment where trunk plugin is not configured, and see the API error as expected:
2023-05-08T20:35:11.179365006+03:00 stderr F E0508 17:35:11.179178       1 instances.go:734] Failed to list trunks: Resource not found    
We will change the code to ignore the error as "no trunk" is a valid configuration.

dulek replied:
I still think a better way would be to check for `trunk` extension on startup and cache that.

stephenfin commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +721,91 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+		klog.Errorf("Failed to list trunks: %v", err)
+		return interfaces, err
+	}
+	allTrunks, err := trunks.ExtractTrunks(allPages)
```
As an aside, neutron has exceptionally good filtering capabilities and the ability to request specific fields. I don't know if that's exposed by gophercloud but we should probably be using that. It's used extensively in nova for performance reasons.

jingczhang replied:
Good point. I will look into filtering with parent ports.

jingczhang replied:
> This is pulling back all trunks without filtering. I don't think this is a safe thing to do in general.

Thank you very much for all the detailed comments. I will try to update the PR to address the set of comments related to this point first.

jingczhang replied:
I have updated the PR yesterday taking care of comments for getSubInterfaces() function, please review.
Re-arrange getSubInterface() per review comments:
Check for trunks attached to VM interfaces only
Collapse two loop (trunks and subports) into one loop

jingczhang replied:
Next PR update is planned for (1) additional code comments (2) adding unit test, plus we will confirm code transparence when trunk plugin is not enabled in neutron

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +721,91 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
 	return nodeAddresses(&srv.Server, interfaces, networkingOpts)
 }
 
+// getSubInterfaces
```
comment taken, will add short description of the function.

jingczhang commented:
> and do you think we can consider add a functional test to this?

Yes, we will look into adding test case.

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +721,91 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+	for _, trunk := range allTrunks {
+		for _, iface := range interfaces {
+			if iface.PortID == trunk.PortID {
+				s, err := trunks.GetSubports(network, trunk.ID).Extract()
```
Acked, will handle this with the comment before.

mdbooth reviewed (changes requested):
This definitely needs unit tests and comments. It's very hard to understand what it's doing.

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -735,68 +735,67 @@ func getAddressesByName(compute *gophercloud.ServiceClient, name types.NodeName,
-				IPAddress: ip.IPAddress,
+		// Get subports attached to the trunks
+		for _, trunk := range allTrunks {
+			subports, err := trunks.GetSubports(network, trunk.ID).Extract()
```
you are right, point taken

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,40 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
 		return nil, err
 	}
 
+	// add subports if exist to the server
+	extraPrivates := make(map[string][]Address)
+	for k, v := range allPrivates {
+		ok := false
+		for _, a := range v {
+			for _, v1 := range addresses {
+				for _, a1 := range v1 {
+					if a.Addr == a1.Addr {
+						ok = true
+						break
+					}
+				}
+			}
+		}
+		if !ok {
+			extraPrivates[k] = v
+		}
+	}
+	klog.V(5).Infof("Node '%s' extraPrivates '%s'", srv.Name, extraPrivates)
+	for k, v := range extraPrivates {
+		v1, ok := addresses[k]
+		if !ok {
+			addresses[k] = v
+		} else {
+			addresses[k] = append(v1, v...)
+		}
+	}
```
This piece of code is to work with the initial version of getSubInterfaces(), i.e., to filter out the subports not used by the VM. It is no longer needed, I will remove it.

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,40 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
 		return nil, err
 	}
 
+	// add subports if exist to the server
```
Related to the above comment, yes, you are absolutely right, the work has been done in the modified getSubInterfaces(). no longer needed, will be removed.

jingczhang commented:
Unit test TestNodeAddressesWithSubports() is added, it will be in the next PR update. I will try to upload the update soon.

jingczhang commented:
PR is updated to address all the comments received so far. Please kindly review to make sure.
Jing

jingczhang commented:
Hi @mdbooth, I think the last PR update has addressed all your comments, can you please check?
Hi all reviewers, please let me know if you have additional comments? also, I think I have replied all comments, if there is pending comments I haven't addressed (esp not addressed from the review procedure point of view) please let me know.

dulek commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,47 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+				extraPrivates[k] = v
+			}
+		}
+		klog.V(5).Infof("Node '%s' extraPrivates '%s'", srv.Name, extraPrivates)
```
This feels like a leftover from debugging. We should name `extraPrivates` in a meaningful way for logs. How about simply "extra private interfaces '%s'"?

jingczhang replied:
good point, I ack it, will do in the next PR update as for the aboves.

dulek commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+
+	var subports []attachinterfaces.Interface
+	for _, iface := range interfaces {
+		// Check if trunk ports are attached
+		listOpts := trunks.ListOpts{
+			PortID: iface.PortID,
+		}
+		allPages, err := trunks.List(network, listOpts).AllPages()
+		if err != nil {
+			klog.Errorf("Failed to list trunks: %v", err)
+			return subports, nil
+		}
+		allTrunks, err := trunks.ExtractTrunks(allPages)
+		if err != nil {
+			klog.Errorf("Failed to extract trunks: %v", err)
+			return subports, err
+		}
+		// Get subports attached to the trunks
+		for _, trunk := range allTrunks {
```
Would it be easier to use trunk_details instead?

```
{
  "port": {
    <snip>
    "id": "713cc70f-fb49-4f7a-ae33-efb9eab93cb3",
    <snip>
    "trunk_details": {
      "sub_ports": [
        {
          "mac_address": "fa:16:3e:5b:bf:e1",
          "port_id": "0cd81ae6-df5d-47f6-ae14-a7ba706f4f23",
          "segmentation_id": 123,
          "segmentation_type": "vlan"
        }
      ],
      "trunk_id": "c264b0a2-b7d6-49a4-82b7-d1b42862d065"
    },
    <snip>
  }
}
```

At a glance Gophercloud doesn't support it, but we can still implement that support there if so. @pierreprinetti, do you find it feasible? How do we implement optional extension fields for Ports in Gophercloud? I feel like it should be possible by extending the Port structure we use?

mdbooth replied:
I thought about this, but I think it's the same number of API calls? We need to fetch the subports anyway in order to get the network name, so whether we get them from the trunk or the parent port doesn't make any difference, I think.

dulek replied:
Yeah, I think so, we end up with the same number of calls.

dulek commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+					PortState: "ACTIVE",
+					FixedIPs:  []attachinterfaces.FixedIP{},
+					PortID:    p.ID,
+					NetID:     n.Name,
```
Aw, but it's probably correct given that we later filter by network name…?

jingczhang replied:
yes, this is to cope with nodeAddresses() filtering on network name

mdbooth replied:
Incidentally, adding this network name here is the only reason for this additional round of API calls. If we aren't using that network name, which could be checked statically from config, we can omit all of this in favour of getting the rest of the info from trunk details as @dulek suggested above.

Lets not go there with this PR, though. This is good, imho.

dulek reviewed (commented):
I think this looks fairly okay after the round of Matt's comments getting addressed. I'm dropping a few comments, but nothing too serious from my side.

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+			return subports, err
+		}
+		// Get subports attached to the trunks
+		for _, trunk := range allTrunks {
```
good point, I ack it

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+					return subports, err
+				}
+				var sface = attachinterfaces.Interface{
+					PortState: "ACTIVE",
```
yes, I ack it, the port state is irrelevant here

jingczhang replied:
Actually, nodeAddresses() filters on port state ACTIVE, so need to set the port state to align with that.

mdbooth commented:
/assign

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,47 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+		extraPrivates := make(map[string][]Address)
+		// For each private network
+		for k, v := range allPrivates {
+			ok := false
```
Suggest ok -> found, because that's what it means here.

jingczhang replied:
comment taken

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,47 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+			}
+			// All the addresses in the private network are not directly connected
+			// Save the private network
+			if !ok {
```
Then this becomes:

```go
if !found {
```

which is much more readable.

jingczhang replied:
comment taken

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
 
-func getAddressesByName(client *gophercloud.ServiceClient, name types.NodeName, networkingOpts NetworkingOpts) ([]v1.NodeAddress, error) {
-	srv, err := getServerByName(client, name)
+func getAddressesByName(compute *gophercloud.ServiceClient, name types.NodeName, networkingOpts NetworkingOpts, network *gophercloud.ServiceClient) ([]v1.NodeAddress, error) {
```
nit: I'd personally group similar arguments here and both both gophercloud.ServiceClient arguments first.

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+		}
+		// Get subports attached to the trunks
+		for _, trunk := range allTrunks {
+			if len(trunk.Subports) == 0 {
+				continue
+			}
```
This is redundant: if Subports is empty the loop below is already a no-op.

jingczhang replied:
comment taken, it is added to avoid the trace for trunk has no subports

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+
 // getAttachedInterfacesByID returns the node interfaces of the specified instance.
-func getAttachedInterfacesByID(client *gophercloud.ServiceClient, serviceID string) ([]attachinterfaces.Interface, error) {
+func getAttachedInterfacesByID(compute *gophercloud.ServiceClient, serviceID string, network *gophercloud.ServiceClient) ([]attachinterfaces.Interface, error) {
```
nit: Again, I'd group the service clients.

jingczhang replied:
comment taken,

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+		listOpts := trunks.ListOpts{
+			PortID: iface.PortID,
+		}
+		allPages, err := trunks.List(network, listOpts).AllPages()
```
We still need to check if trunks are enabled.

jingczhang replied:
This is already checked, API error is returned if trunk plugin is not enable, the API error stops the flow and returns to caller with no subports.

jingczhang replied:
You are right, thanks for the pointer. The patch is built for trunk plugin is configured (normal case) and trunk is used (vlan-aware VM use case), checking for if trunk plugin is configured was not needed so not considered. It seems this checking won't be necessary after baseline code restructure per @mdbooth.

mdbooth commented on pkg/openstack/instances.go:
```diff
@@ -708,6 +819,10 @@ func getAttachedInterfacesByID(client *gophercloud.ServiceClient, serviceID stri
 		return interfaces, err
 	}
-
+	subInterfaces, err := getSubInterfaces(interfaces, network)
```
I want this to be free if trunks aren't in use, but it's not: it's still adding at least 1 additional API call per attached port (list trunks by port id). This seems like a very high cost for a feature most people aren't going to be using.

I wonder if we do need to refactor all of this to fetch ports by device id instead of attachinterfaces, which would allow us to use trunk details.

jingczhang replied:
not really, the getSubInterfaces() returns upon the first trunk API error.

jingczhang replied:
sorry, reading this again and realizing you refer to CPU cycles spent on checking if trunk is configured, this cost depends on number of VMs and ports per VM

mdbooth replied:
Not so much CPU cycles as API calls. Remember we're running this loop for every node in every cluster every 10 minutes by default. At only 100 nodes that's still every 6 seconds on average. Additional API calls here are *expensive*.

mdbooth commented on pkg/openstack/openstack_test.go:
```diff
@@ -878,6 +878,233 @@ func TestNodeAddressesWithAddressSortOrderOptions(t *testing.T) {
 	}
 }
 
+func TestNodeAddressesWithSubports(t *testing.T) {
```
This isn't really testing the interesting code. In fact, nothing here necessarily relates to subports. We need test coverage for getAttachedInterfacesByID(), e.g.:

* Trunking is disabled
* Trunking is enabled but there are no trunks
* A single port has a single subport
* A single port has multiple subports
* Multiple ports have multiple subports

jingczhang replied:
hmm...I will get back for this.

jingczhang replied:
sure, will get resources to add those unit tests

mdbooth reviewed (commented):
Still many things here, but the biggest one is: this code should not add cost for users who don't need it. Achieving that may require some refactoring.

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,47 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
 		return nil, err
 	}
 
+	// Get the private networks that are not directly connected
```
make sense, will try this in next PR update.

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,47 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+	if len(addresses) < len(allPrivates) {
+		extraPrivates := make(map[string][]Address)
+		// For each private network
+		for k, v := range allPrivates {
```
comment taken, can not disagree.

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,47 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+		for k, v := range allPrivates {
+			ok := false
+			// For each address in the private network
+			for _, a := range v {
```
comment taken

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,47 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+			// For each address in the private network
+			for _, a := range v {
+				// Check if the address is directly connected
+				for _, v1 := range addresses {
```
comment taken

jingczhang commented on pkg/openstack/instances.go:
```diff
@@ -613,17 +634,47 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+			for _, a := range v {
+				// Check if the address is directly connected
+				for _, v1 := range addresses {
+					for _, a1 := range v1 {
```
comment taken

jingczhang commented:
Thanks @mdbooth and @dulek for the code comments. I will address them first. Additional unit tests will follow after that.

jingczhang commented:
PR is updated per last round code comments.
Hi @mdbooth, I need pointer for writing the 1st unit test you've requested, it is not obvious to me how it could be done, please advise.
<===
We need test coverage for getAttachedInterfacesByID(), e.g.:
Trunking is disabled

mdbooth reviewed (commented):
The more I think about this, the more I think we should refactor the surrounding code before doing this. The way you've approached this is probably the only reasonable way to do this with the existing code structure. However, it adds an additional api call per attached interface for every node in your cloud, every 10 minutes, whether you're using subports or not. Given that nobody is currently using subports in this way and most people probably never will, this feels like an unreasonably large cost.

I think we need to fetch ports directly from neutron by deviceID in the first instance. This would also show us subports if there are any, meaning the cost of the feature would only be paid by systems which are using it.

In short:
* Refactor nodeAddresses() as above
* Rewrite this PR to use the refactored nodeAddresses

I think the feature is good, but I would prefer that this PR didn't merge in its current form due to its impact on the majority of users.

jingczhang commented:
Hi @mdbooth, thank you very much for your review and plan. We have the interest to upstream this so as to get rid of the "out of tree" rebase trouble. Looking forward to retry this PR after your code restructure then.

mdbooth commented:
> 

Unfortunately although I think this code could do with a refactor with or without the addition of subports, it's not on my todo list right now 😞

mdbooth commented:
FYI I've added the following:
* [Fetch ports instead of interfaces in CPO](https://github.com/kubernetes/cloud-provider-openstack/pull/2250) (not yet merged)
* [Support trunk_details extension in gophercloud](https://github.com/gophercloud/gophercloud/pull/2625) (merged)

Additionally, @pierreprinetti has been noodling with a gophercloud/utils library to allow us to fetch multiple specified ports in a single API call: https://github.com/gophercloud/utils/pull/190.

With all of these in place, I believe:
* subport support adds no overhead for anybody not using it
* subport support requires only 1 additional API call per nodeAddresses() for anybody who is using it (to fetch all detected subports)

mdbooth commented:
@jingczhang both of the above mentioned changes are now merged, and I've bumped the gophercloud dep to 1.4.0 which includes the trunk_details patch. #2250 has almost certainly caused the merge conflict in this patch so you're probably going to have to re-work it even for your downstream fork. By using trunk_details you should now be able to implement this with zero cost for anybody not using the feature. Are you likely to update the PR?

jingczhang commented:
> @jingczhang both of the above mentioned changes are now merged, and I've bumped the gophercloud dep to 1.4.0 which includes the trunk_details patch. #2250 has almost certainly caused the merge conflict in this patch so you're probably going to have to re-work it even for your downstream fork. By using trunk_details you should now be able to implement this with zero cost for anybody not using the feature. Are you likely to update the PR?

Thank you for the information. Sure, I will update the PR.
